stl/src/atomic.cpp

```diff
@@ -4,33 +4,87 @@
 // implement shared_ptr spin lock
 
 #include <yvals.h>
-
+#include <atomic>
 #include <intrin.h>
-#pragma warning(disable : 4793)
+#include <libloaderapi.h>
+#include <synchapi.h>
 
-_EXTERN_C
+#pragma warning(disable : 4793) // '%s' : function is compiled as native code
 
-// SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
-volatile long _Shared_ptr_flag;
+namespace {
+    // MUTEX FOR shared_ptr ATOMIC OPERATIONS
+    SRWLOCK _Shared_ptr_lock = SRWLOCK_INIT;
 
-_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
-#ifdef _M_ARM
-    while (_InterlockedExchange_acq(&_Shared_ptr_flag, 1)) {
-        __yield();
+#if _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA
+    // SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
+    long _Shared_ptr_flag;
+
+
+    struct _Srw_functions_table {
+        _STD atomic<decltype(&::AcquireSRWLockExclusive)> _Pfn_AcquireSRWLockExclusive{nullptr};
+        _STD atomic<decltype(&::ReleaseSRWLockExclusive)> _Pfn_ReleaseSRWLockExclusive{nullptr};
+    };
+
+    _Srw_functions_table _Table;
+
+    bool _Acuire_srw_functions() {
+        enum class _Shared_ptr_api_level : char {
+            __has_nothing = false,
+            __has_srwlock = true,
+            __not_set,
+        };
+        static _STD atomic<_Shared_ptr_api_level> _Api_level{_Shared_ptr_api_level::__not_set};
+
+        _Shared_ptr_api_level _Current = _Api_level.load(_STD memory_order_acquire);
+        if (_Current == _Shared_ptr_api_level::__not_set) {
+            _Current          = _Shared_ptr_api_level::__has_nothing;
+            HMODULE _Kernel32 = GetModuleHandleW(L"Kernel32.dll");
+            if (_Kernel32 != nullptr) {

```

Comment:
Is this permitted for OneCore? (winapisupp.cpp doesn't seem to think we can call GetModuleHandleW.)

Reply:
It is under `_STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA`